google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java

```diff
@@ -33,6 +103,348 @@
    */
   @Nullable
   CloudProject showDialog(@Nullable CloudProject cloudProject) {
-    return cloudProject;
+    // create dialog wrapper for the form once, create services and subscribe to login changes once.
+    if (dialogWrapper == null) {
+      dialogWrapper = new ProjectSelectionDialogWrapper();
+
+      dialogWrapper.setTitle(GctBundle.getString("cloud.project.selector.dialog.title"));
+      // disabled unless project is selected in the list.
+      dialogWrapper.setOKActionEnabled(false);
+      // install additional IDEA UI.
+      installTableSpeedSearch(projectListTable);
+
+      Stream.of(ProjectManager.getInstance().getOpenProjects())
+          .forEach(
+              project ->
+                  project
+                      .getMessageBus()
+                      .connect(dialogWrapper.getDisposable() /* disconnect once dialog is gone. */)
+                      .subscribe(
+                          GoogleLoginListener.GOOGLE_LOGIN_LISTENER_TOPIC,
+                          this::loadAllProjects));
+    }
+
+    loadAllProjects();
+    setSelectedProject(cloudProject);
+
+    DialogManager.show(dialogWrapper);
+
+    int result = dialogWrapper.getExitCode();
+    return result == DialogWrapper.OK_EXIT_CODE ? getSelectedProject() : null;
+  }
+
+  @VisibleForTesting
+  void setSelectedProject(CloudProject cloudProject) {
+    if (cloudProject != null && !Strings.isNullOrEmpty(cloudProject.googleUsername())) {
+      Optional<CredentialedUser> loggedInUser =
+          Services.getLoginService().getLoggedInUser(cloudProject.googleUsername());
+      if (loggedInUser.isPresent()) {
+        selectedProjectsByAccount.put(loggedInUser.get(), cloudProject.projectName());
```

Comment:
Cloud project name is not unique; project ID is. I just want to make sure the logic here and elsewhere is aware of this. In other words, if this is actually using the cloud project name here, then I think this example usage will have problems (since it may not select the correct project).

Reply:
so in `AppEngineDeploymentConfiguration`, the field is `cloudProjectName`, is it actually name or ID?

Reply:
The main thing is that whenever we are keying on it, we must use the ID. But if we are displaying it (like in the new Cloud Project widget - [name] / [email]) we use the name.

Reply:
sounds good, then let's add  `projectId` to keep the actual ID in `CloudProject` too